### PR TITLE
onedrivegui: init at 1.0.3

### LINKED ISSUE #257520: Package request: onedrivegui
**Project description**

A GUI companion for the excellent onedrive for Linux at https://abraunegg.github.io/

- Management and configuration of multiple OneDrive accounts
- Asynchronous real-time monitoring of multiple OneDrive accounts
- Setup wizard for easy OneDrive profile creation and import
- Auto-sync on GUI startup
- Support for GUI based login process
- System tray (if supported by your desktop environment)
- Start minimized to tray/dock
- Input validation to prevent configuration of incompatible OneDrive client options
- Import and management of Business Shared Folders
- Import and management of SharePoint Shared Libraries
- ToolTips with brief explanation of various OneDrive Client configuration options.
- Prompt for re-sync authorization to prevent unexpected data loss.

**Metadata**

* homepage URL: https://github.com/bpozdena/OneDriveGUI
* source URL: https://github.com/bpozdena/OneDriveGUI
* license: mit, bsd, gpl2+ , ... [GPL-3.0 license](https://github.com/bpozdena/OneDriveGUI/blob/main/LICENSE)
* platforms: unix, linux, darwin, ... Linux

### PR DESCRIPTION
**Pull request by chewblacka**
## Description of changes
onedrivegui init at 1.0.3
OneDriveGUI is a GUI for the linux OneDrive client.
https://github.com/bpozdena/OneDriveGUI

Closes #257520

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/by-name/on/onedrivegui/package.nix
Comment (bertof) [inline pkgs/by-name/on/onedrivegui/package.nix:23]: Use something like 

```nix
builtins.attrValues { inherit (python3Packages) pyside6 requests }
```

File: pkgs/by-name/on/onedrivegui/package.nix
Comment (bertof) [inline pkgs/by-name/on/onedrivegui/package.nix:43]: Requires formatting

File: pkgs/by-name/on/onedrivegui/package.nix
Comment (bertof) [inline pkgs/by-name/on/onedrivegui/package.nix:57]: Usually it's a better idea to only use link (ln) from the $src, this way you're not duplicating the files and you're leaving Nix the handling task.

**Review by bertof — CHANGES_REQUESTED** [review]
Hi, the package seem mostly fine. You're missing an entry in the `all-packages` file, so that it can be imported in nixpkgs.
The package builds correctly and the application seem to run. I don't have a configuration ready to test it thoroughly.

**Comment by chewblacka** [comment]
> Hi, the package seem mostly fine. You're missing an entry in the `all-packages` file, so that it can be imported in nixpkgs. The package builds correctly and the application seem to run. I don't have a configuration ready to test it thoroughly.

Many thanks for the detailed and helpful feedback! This is my first package init and is exactly the kind of valuable feedback I was hoping for. I've incorporated all your suggested changes. Also changed the `pname` to `onedrivegui` (instead of `OneDriveGUI`).

Do I still need to manually change `all-packages`? I wasn't sure what the current convention is for new packages. Looking through recently merged PRs they seem to be all over the shop. Some contain additions to `all-packages`, but most don't. If you could clarify it would be much appreciated!

Many thanks again.

**Comment by bertof** [comment]
Hey, nice to see new contributors and contributions. I didn't knew of the per-name packaging either, I was wandering why it still got loaded into the packages. The package seems fine to me as it is now.

**Comment by chewblacka** [comment]
> Hey, nice to see new contributors and contributions. I didn't knew of the per-name packaging either, I was wandering why it still got loaded into the packages. The package seems fine to me as it is now.

Much appreciated!

File: pkgs/by-name/on/onedrivegui/package.nix
Comment (peterhoeg) [inline pkgs/by-name/on/onedrivegui/package.nix:23]: I suggest the more canonical way to write this:

```nix
  propagatedBuildInputs = [ onedrive ] ++ (with python3Packages; [ pyside6 requests ]);
```

File: pkgs/by-name/on/onedrivegui/package.nix
Comment (peterhoeg) [inline pkgs/by-name/on/onedrivegui/package.nix:54]: This looks very strange - why are you pulling in the source?

File: pkgs/by-name/on/onedrivegui/package.nix
Comment (peterhoeg) [inline pkgs/by-name/on/onedrivegui/package.nix:55]: 2 things:

1. while 644 will work, the store is not writable, so you might as well just use 444
2. `install` has a `-t dir` argument so you don't have to repeat the actual file name

In conclusion:
```sh
install -Dm444 -t $out/share/icons/hicolor/48x48/apps src/resources/images/OneDriveGUI.png
```

File: pkgs/by-name/on/onedrivegui/package.nix
Comment (peterhoeg) [inline pkgs/by-name/on/onedrivegui/package.nix:56]: If upstream uses `OneDriveGUI.py` as the executable name, I suggest you leave it as-is and then do a symlink

**Review by peterhoeg — CHANGES_REQUESTED** [review]

File: pkgs/by-name/on/onedrivegui/package.nix
Comment (chewblacka) [reply pkgs/by-name/on/onedrivegui/package.nix:54]: > This looks very strange - why are you pulling in the source?

The binary basically needs access to the subdirectories `resources` and `ui` which contain various assets. It was a choice between copying them into `$out/bin` or linking them from source. Is there a more canonical way to do this? Many thanks!

Comment (chewblacka) [reply pkgs/by-name/on/onedrivegui/package.nix:53]: Would this be better?

    cp -r $src/src/{resources,ui} $out/bin

**Comment by peterhoeg** [comment]
We definitely do not want anything under $out/bin that’s not an executable.

Instead, copy those to $out/share and then patch up the executable to look in the right spot. I haven’t looked at how the program looks up it’s resources and the stuff in ui, but it might be possible to just wrap the executable and fiddle with PYTHON_PATH in the wrapper.

**Comment by chewblacka** [comment]
> Instead, copy those to $out/share and then patch up the executable to look in the right spot. I haven’t looked at how the program looks up its resources and the stuff in ui, but it might be possible to just wrap the executable and fiddle with PYTHON_PATH in the wrapper.

Many thanks for the suggestions and help. The python script hard-codes the path it uses to look for asset folders, but it does prepend the path with `DIR_PATH` which by default it sets to the location of the binary. I've patched this using `sed` to instead point to `$out/share/OneDriveGUI` where I've placed the resource folders. This seems to work. I hope this is an acceptable solution. Many thanks!

**Comment by chewblacka** [comment]
> We definitely do not want anything under $out/bin that’s not an executable. Instead, copy those to $out/share and then patch up the executable to look in the right spot. I haven’t looked at how the program looks up its resources and the stuff in ui, but it might be possible to just wrap the executable and fiddle with PYTHON_PATH in the wrapper.

This is the line I've added which patches the script with the correct path:

    # Patch OneDriveGUI.py so DIR_PATH poitns to shared files location
    sed -i "s|\(DIR_PATH = \)os.path.*|\1'$out/share/OneDriveGUI'|" src/OneDriveGUI.py

File: pkgs/by-name/on/onedrivegui/package.nix
Comment (chewblacka) [inline pkgs/by-name/on/onedrivegui/package.nix:41]: This allows the `OneDriveGUI.py` binary to find the asset folders in `$out/share/OneDriveGUI`

**Comment by peterhoeg** [comment]
I took the liberty of making a few changes - I suggest having a look at what has been done differently here and compare this to your version.

Feel free to adapt it to your own liking. I have added some additional comments - some of which you may want to remove.

``` nix
{ lib
, python3Packages
, fetchFromGitHub
, writeText
, copyDesktopItems
, makeDesktopItem
, makeWrapper
, onedrive
}:

let
  version = "1.0.3";

  # while we can definitely do cat $out/foo <_EOF for smaller files, it's a lot easier to deal with this outside if/when complexity increases
  setupPy = writeText "setup.py" ''
    from setuptools import setup
    setup(
      name='onedrivegui',
      version='${version}',
      scripts=[
        'src/OneDriveGUI.py',
      ],
    )
  '';

in
python3Packages.buildPythonApplication rec {
  pname = "onedrivegui";
  inherit version;

  src = fetchFromGitHub {
    owner = "bpozdena";
    repo = "OneDriveGUI";
    rev = "v${version}";
    hash = "sha256-HutziAzhIDYP8upNPieL2GNrxPBHUCVs09FFxdSqeBs=";
  };

  # makeWrapper is *technically* not needed as `buildPythonApplication` brings it in but this is
  # more explicit which makes it easier to grok (IMHO)
  nativeBuildInputs = [ copyDesktopItems makeWrapper ];

  propagatedBuildInputs = with python3Packages; [ pyside6 requests ];

  # wrap manually to avoid having a bash script in $out/bin with a .py extension
  dontWrapPythonPrograms = true;

  doCheck = false; # No tests defined

  desktopItems = [
    (makeDesktopItem {
      name = "OneDriveGUI";
      exec = "onedrivegui"; # this is the name of our wrapper
      desktopName = "OneDriveGUI";
      comment = "OneDrive GUI Client";
      type = "Application";
      icon = "OneDriveGUI";
      terminal = false;
      categories = [ "Utility" ];
    })
  ];

  postPatch = ''
    # Patch OneDriveGUI.py so DIR_PATH points to shared files location
    sed -i "s|\(DIR_PATH = \)os.path.*|\1'$out/share/OneDriveGUI'|" src/OneDriveGUI.py
    cp ${setupPy} ${setupPy.name}
  '';

  postInstall = ''
    mkdir -p $out/share/OneDriveGUI
    # we do not need the `ui` directory - only resources
    cp -r src/resources $out/share/OneDriveGUI
    install -Dm444 -t $/out/share/icons/hicolor/48x48/apps src/resources/images/OneDriveGUI.png
    # we put our own executable wrapper in place instead
    rm $out/bin/*

    makeWrapper ${python3Packages.python.interpreter} $out/bin/onedrivegui \
      --prefix PATH : ${lib.makeBinPath [ onedrive ]} \
      --prefix PYTHONPATH : ${python3Packages.makePythonPath (propagatedBuildInputs ++ [(placeholder "out")])} \
      --add-flags $out/lib/${python3Packages.python.libPrefix}/site-packages/OneDriveGUI.py
  '';

  meta = with lib; {
    homepage = "https://github.com/bpozdena/OneDriveGUI";
    description = "A simple GUI for Linux OneDrive Client, with multi-account support";
    license = licenses.gpl3Only;
    maintainers = with maintainers; [ jgarcia ];
    platforms = platforms.linux;
  };
}

```

**Comment by chewblacka** [comment]
> I took the liberty of making a few changes - I suggest having a look at what has been done differently here and compare this to your version.
> 
> Feel free to adapt it to your own liking. I have added some additional comments - some of which you may want to remove.

Many thanks for your detailed changes! That was really good of you to go through it so thoroughly. I've kept your file mostly as is, removed one or two comments. Main change is I added an `-r` flag to `rm -r $out/bin` since `$out/bin` contains a `__pycache__` directory which errors and doesn't delete without the `-r` flag. Many thanks again. Much appreciated!

**Comment by chewblacka** [comment]
> Please, keep the comments in the code. They may help future maintainers/users of the package.

I removed this comment:
```
# while we can definitely do cat $out/foo <_EOF for smaller files, it's a lot easier to deal with this outside if/when complexity increases
```
because it only makes sense in the context of my earlier file.

And I removed this comment:

```
  # makeWrapper is *technically* not needed as `buildPythonApplication` brings it in but this is
  # more explicit which makes it easier to grok (IMHO)
```
because it's useful to me, but I don't think is needed for future maintainers.

File: pkgs/by-name/on/onedrivegui/package.nix
Comment (Janik-Haag) [inline pkgs/by-name/on/onedrivegui/package.nix:61]: you should use `substituteInPlace` here instead of sed

Comment (chewblacka) [reply pkgs/by-name/on/onedrivegui/package.nix]: Many thanks for those helpful suggestions! I have implemented all of them and re-pushed a commit.

**Comment by chewblacka** [comment]
All requested changes have been applied. Package can be unblocked and merged unless there is any further changes that need applying.

**Review by Janik-Haag — APPROVED** [review]
> All requested changes have been applied. Package can be unblocked and merged unless there are any further changes that need applying.

Yes I know and the changes LGTM but I would wait a few days if @peterhoeg still has something

File: pkgs/by-name/on/onedrivegui/package.nix
Comment (peterhoeg) [inline pkgs/by-name/on/onedrivegui/package.nix:29]: Do we know which exact version (3.x) is needed?

File: pkgs/by-name/on/onedrivegui/package.nix
Comment (peterhoeg) [inline pkgs/by-name/on/onedrivegui/package.nix:63]: So @Janik-Haag made a good point about how you *in general* want to use `substituteInPlace` (and friends) rather than invoke sed directly, but in this case you run the risk of potentially substituting something unintended as this call is far less specific than what you did with sed.

I would personally just use something like this (untested):
```
sed -i src/OneDriveGUI.py \
  -e "s@^DIR_PATH =@DIR_PATH = '$out/share/OneDriveGUI@'"

**Comment by peterhoeg** [comment]
I'm currently traveling, so a little slow to reply.

File: pkgs/by-name/on/onedrivegui/package.nix
Comment (chewblacka) [reply pkgs/by-name/on/onedrivegui/package.nix:29]: > Do we know which exact version (3.x) is needed?

No, there is no mention of minimum python version that I could find. I don't know what a sensible default would be so I just went with 3.0. Happy to set to something else.

**Comment by peterhoeg** [comment]
> No, there is no mention of minimum python version that I could find. I don’t know what a sensible default would be so I just went with 3.0. Happy to set to something else.

In that case let’s just drop it - we are building it only with py3 anyway.

There’s been a lot of churn on this PR - thank you for your work and patience!

**Comment by Janik-Haag** [comment]
I'm sorry for the overhead I caused, should have been more careful.

**Review by peterhoeg — APPROVED** [review]

**Comment by peterhoeg** [comment]
Thanks all!

**Pull request merged**
